src/commons/crypto/signing/dispatch/signerrouter.rs

```diff
@@ -0,0 +1,676 @@
+        // a new signer that should be registered.
+        Ok(IdentifyResult::Unidentified)
+    }
+
+    fn is_signer_identified_by_handle(
+        &self,
+        signer_provider: &Arc<SignerProvider>,
+        candidate_handle: &Handle,
+    ) -> Result<IdentifyResult, ErrorString> {
+        let (key_identifier, signer_private_key_id) = match Self::decode_signer_handle(candidate_handle) {
+            Err(err) => {
+                error!(
+                    "Internal error: Signer handle '{}' is invalid: {}",
+                    candidate_handle, *err
+                );
+                return Ok(IdentifyResult::Corrupt);
+            }
+            Ok(res) => res,
+        };
+
+        let handle_name = self.signer_mapper.get_signer_name(candidate_handle)?;
+        let signer_name = signer_provider.get_name().to_string();
+        trace!(
+            "Attempting to identify signer '{}' using identity key stored for signer '{}'",
+            signer_name,
+            handle_name
+        );
+
+        let public_key = match self.signer_mapper.get_signer_public_key(candidate_handle) {
+            Ok(res) => res,
+            Err(err) => {
+                error!(
+                    "Internal error: Identity public key for signer '{}' is invalid: {}",
+                    handle_name, err
+                );
+                return Ok(IdentifyResult::Corrupt);
+            }
+        };
+
+        if public_key.key_identifier() != key_identifier {
+            error!(
+                "Internal error: signer handle '{}' is invalid: key identifier mismatch",
+                candidate_handle
+            );
+            return Ok(IdentifyResult::Corrupt);
+        }
+
+        let challenge = "Krill signer verification challenge".as_bytes();
+        let signature = match signer_provider.sign_registration_challenge(&signer_private_key_id, challenge) {
+            Err(SignerError::SignerUnavailable) => {
+                debug!("Signer '{}' could not be contacted", signer_name);
+                return Ok(IdentifyResult::Unavailable);
+            }
+            Err(SignerError::KeyNotFound) => {
+                debug!(
+                    "Signer '{}' not matched: private key id '{}' not found",
+                    signer_name, signer_private_key_id
+                );
+                return Ok(IdentifyResult::Unidentified);
+            }
+            Err(err) => {
+                error!("Signer '{}' is unusable: {}", signer_name, err);
+                return Ok(IdentifyResult::Unusable);
+            }
+            Ok(res) => res,
+        };
+
+        if public_key.verify(challenge, &signature).is_ok() {
+            debug!("Signer '{}' is ready and known, binding", signer_name);
+            let signer_info = signer_provider.get_info().unwrap_or("No signer info".to_string());
+
+            // Drop the read lock so that we can acquire a write lock
```

Comment:
This does not seem to do anything - clippy tells me that this is dropping a reference only. Furthermore it used again straight after this... so this feels a bit wrong.

Generally speaking I prefer to use scoping when it comes to locks.. if you acquire a lock in a function and want to release it later within the same function then you can wrap part of the function in a scoped block to enforce this.

Reply:
Without the drop you get a deadlock. I also prefer blocks to scope lock acquisition and release, I'll look at the code again but in this case if I recall correctly it was proving a bit of pain to achieve the desired result in a readable way using block scopes.

Reply:
Ah. This code was changed, it used to use a RwLock but that was removed when &mut was dropped from the Signer trait by a new release of the rpki-rs crate. The drop is indeed now redundant!